feat(scm): Add Bitbucket Cloud SCM

ushiradineth opened the pull request:
Fixes #179 

Adds Bitbucket Cloud integration as an SCM provider

zachaller commented:
Will also mentioned that you will probably want to enable [webhook](https://github.com/argoproj-labs/gitops-promoter/blob/main/internal/webhookreceiver/server.go) support as well, it's not technically required but it makes the user experience so much better that I feel it is really should be required.

ushiradineth commented:
Thanks @crenshaw-dev! I've synced the PR with the main branch and the checks have succeeded now.

ushiradineth commented:
> Will also mentioned that you will probably want to enable [webhook](https://github.com/argoproj-labs/gitops-promoter/blob/main/internal/webhookreceiver/server.go) support as well, it's not technically required but it makes the user experience so much better that I feel it is really should be required.

@zachaller I did not notice this haha :D. I will go ahead implement this as well.

ushiradineth commented:
@crenshaw-dev @zachaller I have added webhook support and checks are passing. 

Additionally, I renamed bitbucket to bitbucket_cloud (and other instances) incase of a future Bitbucket Server intergration.

crenshaw-dev commented on internal/controller/pullrequest_controller.go:
```diff
@@ -285,6 +286,8 @@ func (r *PullRequestReconciler) getPullRequestProvider(ctx context.Context, pr p
 		return github.NewGithubPullRequestProvider(ctx, r.Client, scmProvider, *secret, gitRepository.Spec.GitHub.Owner) //nolint:wrapcheck
 	case scmProvider.GetSpec().GitLab != nil:
 		return gitlab.NewGitlabPullRequestProvider(r.Client, *secret, scmProvider.GetSpec().GitLab.Domain) //nolint:wrapcheck
+	case scmProvider.GetSpec().Bitbucket != nil:
```
Should we go ahead and name it BitbucketCloud to disambiguate the implementations from the start?

crenshaw-dev commented on docs/getting-started.md:
```diff
@@ -220,9 +219,108 @@ spec:
+5. Give it a name (e.g., "GitOps Promoter")
+6. Select the following permissions:
+   * **Repositories**: Read and Write
+   * **Pull requests**: Read and Write
```
No specific permission for commit statuses?

crenshaw-dev commented on internal/scms/bitbucket_cloud/commit_status.go:
```diff
@@ -0,0 +1,114 @@
+	commitStatus.Status.Sha = commitStatus.Spec.Sha
+
+	// Bitbucket doesn't return an ID for commit statuses, use key as identifier
+	commitStatus.Status.Id = commitStatus.Spec.Name
```
@zachaller any concerns here? Seems maybe as good as any option?

zachaller replied:
It does look like bitbucket returns a uuid

```
{
  "type": "<string>",
  "links": {
    "self": {
      "href": "<string>",
      "name": "<string>"
    },
    "commit": {
      "href": "<string>",
      "name": "<string>"
    }
  },
  "uuid": "<string>",
  "key": "<string>",
  "refname": "<string>",
  "url": "<string>",
  "state": "FAILED",
  "name": "<string>",
  "description": "<string>",
  "created_on": "<string>",
  "updated_on": "<string>"
}
```

ushiradineth replied:
Sorry for overlooking this, I have fixed this now.

crenshaw-dev commented on internal/scms/bitbucket_cloud/pullrequest.go:
```diff
@@ -0,0 +1,390 @@
+				statusCode = http.StatusInternalServerError
+			}
+		}
+	}
```
Share helper func across similar uses?

emirot commented on internal/scms/bitbucket_cloud/commit_status.go:
```diff
@@ -0,0 +1,102 @@
+	commitStatusOptions := &bitbucket.CommitStatusOptions{
+		State:       phaseToBuildState(commitStatus.Spec.Phase),
+		Key:         commitStatus.Spec.Name,
+		Url:         commitStatus.Spec.Url,
```
@ushiradineth 
From what I saw testing it looks like the URL is empty and bitbucket does not allow that it is a required parameter
@zachaller does it need to be the url or having `example.org` will work just fine?

emirot replied:
@ushiradineth Actually an empty URL won't work we will also need https://github.com/ushiradineth/gitops-promoter/pull/5/files

zachaller commented on internal/scms/bitbucket_cloud/utils.go:
```diff
@@ -0,0 +1,88 @@
+}
+
+// TruncateString safely truncates a string to a maximum number of runes (characters).
+func truncateString(str string, length int) string {
```
```
func truncateString(s string, maxRunes int) string {
	runes := []rune(s)
	if len(runes) <= maxRunes {
		return s
	}
	return string(runes[:maxRunes])
}
```

zachaller replied:
There is already a TruncateString function

emirot commented:
> @emirot Thank you for the fixes, I've updated the PR with latest main and [ushiradineth/gitops-promoter#5 (files)](https://github.com/ushiradineth/gitops-promoter/pull/5/files)

Oh it looks like there is few lint issues sorry about that

crenshaw-dev commented on api/v1alpha1/common_types.go:
```diff
@@ -91,6 +94,18 @@ type ForgejoRepo struct {
+	Workspace string `json:"workspace"`
+	// Repository is the repository of the repository.
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:MinLength=1
```
Unlike owner, turns out the max length of this is actually well documented: https://support.atlassian.com/bitbucket-cloud/docs/create-a-repository/

Based on some light testing, I think the given regex should cover the possible characters. We can adjust later if it's too restrictive.

```suggestion
	// +kubebuilder:validation:MaxLength=62
	// +kubebuilder:validation:Pattern="^[a-zA-Z0-9_-]+.$"
```

crenshaw-dev commented on api/v1alpha1/common_types.go:
```diff
@@ -91,6 +94,20 @@ type ForgejoRepo struct {
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:MaxLength=62
+	// +kubebuilder:validation:Pattern="^[a-zA-Z0-9_-]+.$"
+	Repository string `json:"repository"`
```
For consistency with the other providers, should we call this Name?

ushiradineth replied:
Resolved!

ushiradineth commented:
@crenshaw-dev Yikes sorry about that, removed the nix files, last commit should be clean now
